cynet/client.go

```diff
@@ -0,0 +1,582 @@
+			}
+			// Retry the request with new token
+			retryCount++
+			continue
+
+		case http.StatusForbidden:
+			// Access forbidden - shut down adapter
+			a.conf.ClientOptions.OnError(fmt.Errorf("received 403 Forbidden - access denied to alerts API"))
+			go a.Close()
+			return nil, fmt.Errorf("access forbidden (403) - shutting down adapter")
+
+		case http.StatusTooManyRequests:
+			// Handle rate limiting
+			retryAfter := resp.Header.Get("Retry-After")
+			waitDuration := 60 * time.Second // Default wait
+
+			if retryAfter != "" {
+				if seconds, err := strconv.Atoi(retryAfter); err == nil {
+					waitDuration = time.Duration(seconds) * time.Second
+				}
+			}
+
+			a.conf.ClientOptions.OnWarning(fmt.Sprintf("rate limited (429), waiting %v before retry", waitDuration))
+
+			select {
+			case <-a.ctx.Done():
+				return nil, a.ctx.Err()
+			case <-time.After(waitDuration):
+				retryCount++
+				continue
+			}
+
+		case http.StatusInternalServerError, http.StatusBadGateway, http.StatusServiceUnavailable:
+			// Server errors - retry with exponential backoff
+			lastErr = fmt.Errorf("server error %d: %s", resp.StatusCode, string(body))
+			a.conf.ClientOptions.DebugLog(lastErr.Error())
+			retryCount++
+
+			if retryCount < maxRetries {
+				backoff := time.Duration(retryCount*retryCount) * time.Second
+				a.conf.ClientOptions.OnWarning(fmt.Sprintf("server error %d, retrying in %v (attempt %d/%d)",
+					resp.StatusCode, backoff, retryCount, maxRetries))
+
+				select {
+				case <-a.ctx.Done():
+					return nil, a.ctx.Err()
+				case <-time.After(backoff):
+					continue
+				}
+			}
+
+		default:
+			return nil, fmt.Errorf("unexpected status code %d: %s", resp.StatusCode, string(body))
+		}
+	}
+
+	if lastErr != nil {
+		return nil, fmt.Errorf("request failed after %d retries: %w", maxRetries, lastErr)
+	}
+
+	return nil, fmt.Errorf("request failed after %d retries", maxRetries)
+}
+
+func (a *CynetAdapter) processAlerts(resp *AlertsResponse) ([]utils.Dict, time.Time) {
+	var newAlerts []utils.Dict
+	var latestTime time.Time
+
+	// Parse the SyncTimeUtc
+	syncTime, err := time.Parse(time.RFC3339Nano, resp.SyncTimeUtc)
+	if err != nil {
+		a.conf.ClientOptions.OnError(fmt.Errorf("failed to parse SyncTimeUtc '%s': %v", resp.SyncTimeUtc, err))
+		syncTime = a.since
+	}
+
+	for _, entity := range resp.Entities {
+		// Generate hash ID for the entire entity
+		a.fnvHasher.Reset()
+		b, err := json.Marshal(entity)
+		if err != nil {
+			a.conf.ClientOptions.OnError(fmt.Errorf("failed to marshal alert entity for hashing: %v", err))
+			continue
+		}
+		a.fnvHasher.Write(b)
+		entityID := fmt.Sprintf("%d", a.fnvHasher.Sum64())
+
+		// Check if we've seen this exact entity before
+		if _, seen := a.alertsDedupe[entityID]; seen {
+			continue
+		}
+
+		entity["event-type"] = "cynet-alert"
+
+		// Track for deduplication using sync time
+		a.alertsDedupe[entityID] = time.Now().Unix()
+
+		newAlerts = append(newAlerts, entity)
+
```

Comment:
Using `time.Now()` loses the original event timestamp. Consider extracting timestamp from the alert data if available (e.g., from a `timestamp`, `created_at`, or similar field in the entity).